### ISSUE TITLE
Unacked messages when using SimpleMessageListenerContainer

### ISSUE THREAD

**Issue by novtor** [spring-projects/spring-amqp#694]
Hello,

We are experiencing an issue with `SimpleMessageListenerContainer`. Sometimes we have messages that stay unacked until we restart our service or we kill the connection via RabbitMQ console.

It happens rarely. I got the thread dump and heap dump of a service when it contains unacked messages. From the thread dump I cannot see anything blocked:

```
AMQP Connection 127.0.0.1:5672" Id=274746 in RUNNABLE (running in native)
at java.net.SocketInputStream.socketRead0(Native Method)
at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
at java.net.SocketInputStream.read(SocketInputStream.java:170)
at java.net.SocketInputStream.read(SocketInputStream.java:141)
at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
at java.io.BufferedInputStream.read(BufferedInputStream.java:265)
  - locked java.io.BufferedInputStream@41712376
at java.io.DataInputStream.readUnsignedByte(DataInputStream.java:288)
at com.rabbitmq.client.impl.Frame.readFrom(Frame.java:91)
at com.rabbitmq.client.impl.SocketFrameHandler.readFrame(SocketFrameHandler.java:164)
  - locked java.io.DataInputStream@3509225a
at com.rabbitmq.client.impl.AMQConnection$MainLoop.run(AMQConnection.java:571)
at java.lang.Thread.run(Thread.java:745)
```
and

```
pool-2-thread-28084" Id=274747 in TIMED_WAITING on lock=java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@1c4066eb
at sun.misc.Unsafe.park(Native Method)
at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:1093)
at java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.take(ScheduledThreadPoolExecutor.java:809)
at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)
```

And in the heap dump I can see that the `BlockingQueueConsumer` containing the queue (of type `LinkedBlockingQueue`) is not referenced by any `SimpleMessageListenerContainer` as it should be as all our RabbitMQ listeners are configured to use it.

From the `SimpleMessageListenerContainer` code I saw that it may remove idle consumers. My question is: do you have any idea why it could remove it and not cancel the messages it contains?

RabbitMQ 3.6.4, Erlang 19.1
Spring amqp:spring-rabbit:1.7.3.RELEASE
com.rabbit:amqp-client:4.0.2

here is how we create the factory:

```
 SimpleRabbitListenerContainerFactory factory = new   SimpleRabbitListenerContainerFactory();
        factory.setConnectionFactory(connectionFactory);
        factory.setConcurrentConsumers(min);
        factory.setMaxConcurrentConsumers(max);
        factory.setAcknowledgeMode(AUTO);
        factory.setDefaultRequeueRejected(false);
        factory.setMessageConverter(jacksonMessageConverter);
        // Add an incoming message logging interceptor
        factory.setAdviceChain(
                messageLoggingInterceptor,
                RetryInterceptorBuilder
                        .stateless()
                        .retryPolicy(new NeverRetryPolicy())
                        .recoverer(messageRecoverer)             // to republish failed messages in a dlq
                        .build(),
                rabbitConsumerExceptionLoggingInterceptor
        );
        return factory;
```
       

and the connection factory:

```
com.rabbitmq.client.ConnectionFactory clientFactory = new com.rabbitmq.client.ConnectionFactory();
            if (this.useSsl) {
                clientFactory.useSslProtocol();
            }
            connectionFactory = new CachingConnectionFactory(clientFactory);
            connectionFactory.setUsername(this.username);
            connectionFactory.setPassword(this.password);
            connectionFactory.setAddresses(this.url);
            connectionFactory.setCacheMode(CacheMode.CONNECTION);
            connectionFactory.setChannelCacheSize(this.maximumChannelsPerConnection);
            connectionFactory.setChannelCheckoutTimeout(this.channelCheckoutTimeout);
            connectionFactory.setConnectionCacheSize(this.maxConnections);
```
            
I can provide you with a sample if it looks too messy here but anyway it is difficult to reproduce. The only correlation we saw is that it happens sometimes on queues that idle since some time, that is, no messages during few hours

Thank you

**Comment by artembilan** [comment]
Would you mind to upgrade to recently released Spring AMQP `1.7.5`?
There are some fixes in the `SimpleMessageListenerContainer` around consumers.

I wonder if your problem is covered with them as well.

Thanks for understanding

**Comment by garyrussell** [comment]
Yes, there was a race condition with auto-recovered channels. This was fixed in 1.7.5.

If you can't upgrade for some reason, turn off `automaticRecoveryEnabled` in the `clientFactory` - Spring AMQP doesn't need it - it has had its own recovery mechanisms since day one.

Closing this for now; if you still experience a problem after upgrading (or turning off auto recovery) please reopen.

**Comment by jdathor** [comment]
Hi, 

Just found this thread today, and I think we have the same issue - "Sometimes we have messages that stay unacked until we restart our service or we kill the connection via RabbitMQ console."
But the difference is I have consumer in RabbitMQ greater than the number of active consumer in PCF.

Here is the link in stackoverflow I just created.
https://stackoverflow.com/questions/63465726/pcf-rabbitmq-unacked-messages-in-queue

RabbitMQ 3.8.5
Erlang 22.3.4.3
spring-amqp:2.0.5.RELEASE
amqp-client:5.1.2


Thank you
